internal/mesh/internal/controllers/implicitdestinations/controller.go

```diff
@@ -0,0 +1,314 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package implicitdestinations
+
+import (
+	"context"
+	"sort"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/controller/cache"
+	"github.com/hashicorp/consul/internal/controller/cache/index"
+	"github.com/hashicorp/consul/internal/controller/dependency"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/internal/storage"
+	pbauth "github.com/hashicorp/consul/proto-public/pbauth/v2beta1"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v2beta1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// Future work: this can be optimized to omit:
+//
+// - destinations denied due to DENY TP
+// - only ports exposed by CR or CTP explicitly
+
+/*
+Data Relationships:
+
+Reconcile:
+- read WI[source] (ignore)
+- list CTPs by WI[source]
+  - turn CTP.id -> WI[backend].id
+  - list SVCs by WI[backend]
+    - list CRs by SVC[backend]
+      - turn CR.id -> SVC[dest].id
+	  - emit SVC[dest]
+
+DepMappers:
+- CR <list> SVC[backend] <list> WI[backend] <align> CTP <list> WI[source] <align> CID
+-           SVC[backend] <list> WI[backend] <align> CTP <list> WI[source] <align> CID
+-                                                   CTP <list> WI[source] <align> CID
+- bound refs for all
+
+*/
+
+func Controller(globalDefaultAllow bool) *controller.Controller {
+	m := &mapAndTransformer{globalDefaultAllow: globalDefaultAllow}
+
+	boundRefsMapper := dependency.CacheListMapper(pbmesh.ComputedImplicitDestinationsType, boundRefsIndex.Name())
+
+	return controller.NewController(ControllerID,
```

Comment:
In general the code was extracted like:
- `index.go` : general cache index stateless helpers and the actual `*index.Index` structures
- `mapper.go` : DependencyMapper, transformers, and associated stateful helpers to assist in mapping chains